### PR TITLE
[#40] Support date formats of the form '29-08-2013 02:00:00 CEST'

### PR DESCRIPTION
**Pull request by jonase**

### CODE REVIEW

File: api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj [hunk 1]
```diff
@@ -103,9 +104,14 @@
   [_ response-str]
   response-str)
 
+(def ^DateFormat sdf (SimpleDateFormat. "dd-MM-yyyy HH:mm:ss z"))
```
Comment (iperdomo) [inline api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: `SimpleDateFormat` is not thread safe. The usual way to _solve_ the problem is put it in a ThreadLocal.
https://github.com/akvo/akvo-flow/blob/v1.9.15/GAE/src/org/waterforpeople/mapping/app/web/DataBackoutServlet.java#L71-L76

Can this affect us?

Comment (iperdomo) [reply api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: I would change the var name to `date-format`

Comment (iperdomo) [reply api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: More useful background info on the issues behind `SimpleDateFormat` -  http://stackoverflow.com/a/4021932

Comment (iperdomo) [reply api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: Not sure if i really understand this explanation - https://clojure.org/reference/vars#set

Comment (iperdomo) [reply api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: It seems that declaring the var as `:^dynamic`, and using `(binding [date-format (SimpleDateFormat. "dd-MM-yyyy HH:mm:ss z")] ...)` or just declare the new SimpleDateFormat object inside a `let` ?

File: api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj [hunk 2]
```diff
@@ -103,9 +104,14 @@
...
+
 (defmethod parse-response "DATE"
   [_ response-str]
-  (ds/to-iso-8601 (java.util.Date. (Long/parseLong response-str))))
+  (let [date (try (java.util.Date. (Long/parseLong response-str))
+                  (catch NumberFormatException e
+                    (.parse sdf response-str)))]
```
Comment (iperdomo) [inline api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: this `parse` can also throw, right? https://docs.oracle.com/javase/8/docs/api/java/text/DateFormat.html#parse-java.lang.String-

File: api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj (hunk 1 above)
Comment (jonase) [reply api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: Good catch!

I think we have two options:

```
(locking sdf
    (.parse sdf response-str))
```

or create a new one every time
```
(let [date-format (SimpleDateFormat. "dd-MM-yyyy HH:mm:ss z")]
  (parse date-format response-str))
```

File: api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj (hunk 2 above)
Comment (jonase) [reply api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: yes, we can get a `ParseException`. I don't want to swallow it and return `nil` at this point. I think it's better to wait for one and see which formats are out there.

Especially now with the quick turn-around. We get a message in sentry within a few seconds. The fix is quick and deployment automatic :)

Comment (iperdomo) [reply api/src/clojure/org/akvo/flow_api/datastore/form_instance.clj]: :+1:

**Pull request merged**